### PR TITLE
Improve config and command line arguments

### PR DESCRIPTION
**Pull request by aaronbojarski**
This is an attempt to improve the config and command line arguments.

- the config file gets loaded once and the svcConfig struct is returned
- several functions use this cfg struct to extract the necessary data
- this also adds some tool entries for nts to the existing config entries

Some open issues

- [ ] the naming of the data extracting functions is not consistent
- [ ] is the log argument necessary in the config functions

### CODE REVIEW

File: timeservice.go [hunk 1]
```diff
@@ -184,98 +227,130 @@ func newDaemonConnector(ctx context.Context, log *zap.Logger, daemonAddr string)
...
+	var cfg svcConfig
+	raw, err := os.ReadFile(configFile)
+	if err != nil {
+		log.Fatal("failed to load configuration", zap.Error(err))
+	}
+	err = toml.NewDecoder(bytes.NewReader(raw)).DisallowUnknownFields().Decode(&cfg)
+	if err != nil {
+		log.Fatal("failed to decode configuration", zap.Error(err))
+	}
+
```
Comment (marcfrei) [inline timeservice.go:240]: Remove empty line.

File: timeservice.go [hunk 2]
```diff
@@ -184,98 +227,130 @@ func newDaemonConnector(ctx context.Context, log *zap.Logger, daemonAddr string)
 	return c
 }
 
-func loadConfig(ctx context.Context, log *zap.Logger,
-	configFile, daemonAddr string, localAddr *snet.UDPAddr) (
+func loadConfig(log *zap.Logger, configFile string) svcConfig {
+	var cfg svcConfig
```
Comment (marcfrei) [inline timeservice.go:231]: Move declaration above line 236

File: timeservice.go [hunk 3]
```diff
@@ -184,98 +227,130 @@ func newDaemonConnector(ctx context.Context, log *zap.Logger, daemonAddr string)
...
+func remoteAddress(log *zap.Logger, cfg svcConfig) *snet.UDPAddr {
+	if cfg.RemoteAddress == "" {
+		log.Fatal("remote_address not specified in config")
+	}
+	var remoteAddr snet.UDPAddr
+	err := remoteAddr.Set(cfg.RemoteAddress)
+	if err != nil {
+		log.Fatal("failed to parse remote address")
+	}
+
```
Comment (marcfrei) [inline timeservice.go:266]: Remove empty line.

File: timeservice.go [hunk 4]
```diff
@@ -53,12 +53,17 @@ const (
...
 type svcConfig struct {
-	MBGReferenceClocks []string `toml:"mbg_reference_clocks,omitempty"`
-	NTPReferenceClocks []string `toml:"ntp_reference_clocks,omitempty"`
-	SCIONPeers         []string `toml:"scion_peers,omitempty"`
-	NTSKECertFile      string   `toml:"ntske_cert_file,omitempty"`
-	NTSKEKeyFile       string   `toml:"ntske_key_file,omitempty"`
-	NTSKEServerName    string   `toml:"ntske_server_name,omitempty"`
+	LocalAddress            string   `toml:"local_address,omitempty"`
+	DaemonAddress           string   `toml:"daemon_address,omitempty"`
+	RemoteAddress           string   `toml:"remote_address,omitempty"`
```
Comment (marcfrei) [inline timeservice.go:58]: Just "Addr" instead of "Address".

File: timeservice.go [hunk 5]
```diff
@@ -184,98 +227,130 @@ func newDaemonConnector(ctx context.Context, log *zap.Logger, daemonAddr string)
...
-	raw, err := os.ReadFile(configFile)
-	if err != nil {
-		log.Fatal("failed to load configuration", zap.Error(err))
-	}
-	err = toml.NewDecoder(bytes.NewReader(raw)).DisallowUnknownFields().Decode(&cfg)
-	if err != nil {
-		log.Fatal("failed to decode configuration", zap.Error(err))
-	}
-
+func tlsConfig(log *zap.Logger, cfg svcConfig) *tls.Config {
```
Comment (marcfrei) [inline timeservice.go:353]: I would probably remove the empty lines in this function.

File: timeservice.go [hunk 6]
```diff
@@ -293,10 +368,13 @@ func loadNTSKEConfig(log *zap.Logger, configFile string) *tls.Config {
...
 }
 
-func runServer(configFile, daemonAddr string, localAddr *snet.UDPAddr) {
+func runServer(configFile string) {
 	ctx := context.Background()
 
-	refClocks, netClocks := loadConfig(ctx, log, configFile, daemonAddr, localAddr)
+	cfg := loadConfig(log, configFile)
+	localAddr := localAddress(log, cfg)
+	daemonAddr := cfg.DaemonAddress
```
Comment (marcfrei) [inline timeservice.go:376]: Add a ( trivial) function `deamonAddresss(log *zap.Logger, cfg svcConfig)...`, just for consistency?

**Review by marcfrei — CHANGES_REQUESTED** [review]

File: core/client/client_scion.go [hunk 7]
```diff
@@ -148,6 +148,9 @@ func (c *SCIONClient) measureClockOffsetSCION(ctx context.Context, log *zap.Logg
 			return offset, weight, err
 		}
 		remoteAddr.Host.IP = net.ParseIP(ntskeData.Server)
+		if remoteAddr.Host.IP.To4() != nil {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply core/client/client_scion.go]: Done

File: testnet/test-server.toml [hunk 8]
```diff
@@ -1,4 +1,5 @@
 ntp_reference_clocks = ["0-0,time.facebook.com:123"]
+local_address = "0-0,0.0.0.0:10123"
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply testnet/test-server.toml:1]: Done

But need to decide what should be default address to make sure that the server can access reference clocks and give correct address to client during ntske

File: timeservice.go [hunk 9]
```diff
@@ -135,14 +140,48 @@ func (c *mbgReferenceClock) MeasureClockOffset(ctx context.Context, log *zap.Log
...
+	c.Auth.NTSKEFetcher.TLSConfig = tls.Config{
+		InsecureSkipVerify: ntskeInsecureSkipVerify,
+		ServerName:         ntskeHost,
+		MinVersion:         tls.VersionTLS13,
+	}
+	c.Auth.NTSKEFetcher.Port = ntskePort
+	c.Auth.NTSKEFetcher.Log = log
+}
+
+func configureSCIONClientNTS(c *client.SCIONClient, ntskeServer string, ntskeInsecureSkipVerify bool) {
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply timeservice.go]: Done

File: timeservice.go (hunk 2 above)
Comment (aaronbojarski) [reply timeservice.go]: Done

File: timeservice.go (hunk 3 above)
Comment (aaronbojarski) [reply timeservice.go]: Done

File: timeservice.go (hunk 5 above)
Comment (aaronbojarski) [reply timeservice.go:350]: Done

File: README.md [hunk 10]
```diff
@@ -34,7 +34,7 @@ git clone https://github.com/marcfrei/scion-time.git
 cd ~/scion-time
 go build timeservice.go timeservicex.go
 
-~/scion-time/timeservice server -verbose -local 0-0,0.0.0.0:10123
+~/scion-time/timeservice server -verbose -config testnet/server.toml
```
[the remark this answers is not in the document]
Comment (aaronbojarski) [reply README.md:37]: I like having this as a command line option. During benchmarking it was convenient to test if the config is correct with the `-verbose` flag. I could then run almost the same command again without having to change the config file.

But I would be happy to hear your thoughts.

File: timeservice.go (hunk 6 above)
Comment (marcfrei) [reply timeservice.go:371]: Tough one. But yes, I think an accessor function would look nicer and more consistent.

File: timeservice.go (hunk 4 above)
Comment (aaronbojarski) [reply timeservice.go]: Done

File: timeservice.go (hunk 6 above)
Comment (aaronbojarski) [reply timeservice.go]: Done

**Review by marcfrei — APPROVED** [review]

**Comment by marcfrei** [comment]
Thanks for all the work on this, Aaron.

**Pull request merged**